src/Command/DoPopulateIndex.php

```diff
@@ -0,0 +1,169 @@
+            ->setHidden(true)
+            ->setDescription('[INTERNAL]')
+            ->addOption(self::OPTION_CONFIG, mode: InputOption::VALUE_REQUIRED)
+            ->addOption(self::OPTION_INDEX, mode: InputOption::VALUE_REQUIRED)
+            ->addOption(self::OPTION_BATCH_NUMBER, mode: InputOption::VALUE_REQUIRED)
+            ->addOption(self::OPTION_LISTING_COUNT, mode: InputOption::VALUE_REQUIRED)
+            ->addOption(self::OPTION_DOCUMENT, mode: InputOption::VALUE_REQUIRED)
+        ;
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $indexConfig = $this->getIndex();
+
+        if (!$indexConfig instanceof IndexInterface) {
+            return self::FAILURE;
+        }
+
+        return $this->populateIndex($indexConfig, $indexConfig->getBlueGreenInactiveElasticaIndex());
+    }
+
+    private function getIndex(): ?IndexInterface
+    {
+        foreach ($this->indexRepository->flattenedAll() as $indexConfig) {
+            if ($indexConfig->getName() === $this->input->getOption(self::OPTION_CONFIG)) {
+                return $indexConfig;
+            }
+        }
+
+        return null;
+    }
+
+    private function populateIndex(IndexInterface $indexConfig, ElasticaIndex $esIndex): int
+    {
+        ProgressBar::setFormatDefinition('custom', "%percent%%\t%remaining%\t%memory%\n%message%");
+
+        $batchNumber = (int) $this->input->getOption(self::OPTION_BATCH_NUMBER);
+        $listingCount = (int) $this->input->getOption(self::OPTION_LISTING_COUNT);
+
+        $allowedDocuments = $indexConfig->getAllowedDocuments();
+        $document = $this->input->getOption(self::OPTION_DOCUMENT);
+
+        if (!in_array($document, $allowedDocuments, true)) {
+            return self::FAILURE;
+        }
+
+        $progressBar = new ProgressBar($this->output, $listingCount > 0 ? $listingCount : 1);
+        $progressBar->setMessage($document);
+        $progressBar->setFormat('custom');
+        $progressBar->setProgress($batchNumber * $indexConfig->getBatchSize());
+
+        if (!$indexConfig->shouldIndexInSubprocesses()) {
+            $numberOfBatches = ceil($listingCount / $indexConfig->getBatchSize());
+
+            for ($batch = 0; $batch < $numberOfBatches; $batch++) {
+                $exitCode = $this->doPopulateIndex($esIndex, $indexConfig, $progressBar, $document, $batch);
+
+                if ($exitCode !== self::SUCCESS) {
+                    return $exitCode;
+                }
+            }
+        } else {
+            return $this->doPopulateIndex($esIndex, $indexConfig, $progressBar, $document, $batchNumber);
+        }
+
+        return self::SUCCESS;
+    }
+
+    private function doPopulateIndex(
+        ElasticaIndex $esIndex,
+        IndexInterface $indexConfig,
+        ProgressBar $progressBar,
+        string $document,
+        int $batchNumber,
+    ): int {
+        $documentInstance = $this->documentRepository->get($document);
```

Comment:
Can you please check the original implementation regarding un-/setting the tenant? At least one call to `$this->documentHelper->setTenantIfNeeded($documentInstance, $indexConfig);` is missing here as indexing with tenants crashes:

```
$  console val:el:i -p product_xxx
Index: product_xxx
-> Ensured indices are correctly set up with alias
-> Re-created inactive blue/green index

For large indexes please consider to implement `shouldIndexInSubprocesses` to prevent memory exhaustion.

-> Populating index App\Elasticsearch\Index\Product\ProductIndex batch 1/1.

0%      < 1 sec 18.0 MiB
App\Elasticsearch\Index\Product\ProductIndexDocument
Error while populating index App\Elasticsearch\Index\Product\ProductIndex, processing documents of type App\Elasticsearch\Index\Product\ProductIndexDocument, last processed element ID 9451.

In /var/www/html/packages-local/pimcore-elastica-bridge/src/Document/AbstractTenantAwareDocument.php line 29

Tenant has not yet been set

#0 /var/www/html/src/Elasticsearch/Document/AbstractTenantAwareIndexDocument.php(55): Valantic\ElasticaBridgeBundle\Document\AbstractTenantAwareDocument->getTenant()
#1 /var/www/html/src/Elasticsearch/Index/Product/ProductIndexDocument.php(77): App\Elasticsearch\Document\AbstractTenantAwareIndexDocument->getTenantEnum()
#2 /var/www/html/packages-local/pimcore-elastica-bridge/src/Command/DoPopulateIndex.php(128): App\Elasticsearch\Index\Product\ProductIndexDocument->shouldIndex(Object(Pimcore\Model\DataObject\Product))
#3 /var/www/html/packages-local/pimcore-elastica-bridge/src/Command/DoPopulateIndex.php(96): Valantic\ElasticaBridgeBundle\Command\DoPopulateIndex->doPopulateIndex(Object(Elastica\Index), Object(App\Elasticsearch\Index\Product\ProductIndex), Object(Symfony\Component\Console\Helper\ProgressBar), 'App\\Elasticsear...', 0)
#4 /var/www/html/packages-local/pimcore-elastica-bridge/src/Command/DoPopulateIndex.php(58): Valantic\ElasticaBridgeBundle\Command\DoPopulateIndex->populateIndex(Object(App\Elasticsearch\Index\Product\ProductIndex), Object(Elastica\Index))
#5 /var/www/html/vendor/symfony/console/Command/Command.php(326): Valantic\ElasticaBridgeBundle\Command\DoPopulateIndex->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#6 /var/www/html/vendor/symfony/console/Application.php(1096): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#7 /var/www/html/vendor/symfony/framework-bundle/Console/Application.php(126): Symfony\Component\Console\Application->doRunCommand(Object(Valantic\ElasticaBridgeBundle\Command\DoPopulateIndex), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#8 /var/www/html/vendor/symfony/console/Application.php(324): Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand(Object(Valantic\ElasticaBridgeBundle\Command\DoPopulateIndex), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#9 /var/www/html/vendor/symfony/framework-bundle/Console/Application.php(80): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#10 /var/www/html/vendor/symfony/console/Application.php(175): Symfony\Bundle\FrameworkBundle\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#11 /var/www/html/vendor/symfony/runtime/Runner/Symfony/ConsoleApplicationRunner.php(49): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#12 /var/www/html/vendor/autoload_runtime.php(29): Symfony\Component\Runtime\Runner\Symfony\ConsoleApplicationRunner->run()
#13 /var/www/html/bin/console(20): require_once('/var/www/html/v...')
#14 {main}


In DoPopulateIndex.php line 138:
                            
  Indexing document failed  
                            

In AbstractTenantAwareDocument.php line 29:
                               
  Tenant has not yet been set  

```

Reply:
Still investigating this ...

Reply:
Unfortunately, I'm not able to reproduce this. Of course I can check the original implementation if I'm something missing but would be great to test it. Do you have any advice how I can reproduce this?

Reply:
Should be resolved